qiskit_experiments/library/randomized_benchmarking/sampling_utils.py

```diff
@@ -0,0 +1,359 @@
+    from :math:`E_r`.
+
+    2. Select edges from :math:`E` with the probability :math:`w\xi/2|E|`. These
+    edges will have two-qubit gates in the output layer.
+
+    |
+
+    This produces a layer with an expected two-qubit gate density :math:`\xi`. In the
+    default mirror RB configuration where these layers are dressed with single-qubit
+    Pauli layers, this means the overall expected two-qubit gate density will be
+    :math:`\xi_s/2=\xi`. The actual density will converge to :math:`\xi_s` as the
+    circuit size increases.
+
+    .. ref_arxiv:: 1 2008.11294
+
+    """
+
+    def __call__(
+        self,
+        qubits: Sequence,
+        length: int = 1,
+    ) -> List[Tuple[GateInstruction]]:
+        """Sample layers using the edge grab algorithm.
+
+        Args:
+            qubits: A sequence of qubits to generate layers for.
+            length: The length of the sequence to output.
+            seed: Seed for random generation.
+
+        Raises:
+            Warning: If the coupling map has no connectivity or
+                ``two_qubit_gate_density`` is too high.
+            TypeError: If invalid gate set(s) are specified.
+
+        Returns:
+            A ``length``-long list of :class:`qiskit.circuit.QuantumCircuit` layers over
+            ``num_qubits`` qubits. Each layer is represented by a list of tuples which
+            are in the format ((one or more qubit indices), gate). Single-qubit
+            Cliffords are represented by integers for speed. Here's an example with the
+            default choice of Cliffords for the single-qubit gates and CXs for the
+            two-qubit gates:
+
+            .. parsed-literal::
+                (((1, 2), CXGate()), ((0,), 12), ((3,), 20))
+
+            This represents a layer where the 12th Clifford is performed on qubit 0,
+            a CX is performed with control qubit 1 and target qubit 2, and the 20th
+            Clifford is performed on qubit 3.
+
+        """
+        num_qubits = len(qubits)
+        gateset = self._probs_by_gate_size()
+        norm1q = sum(gateset[1][1])
+        norm2q = sum(gateset[2][1])
+        two_qubit_gate_density = sum(gateset[2][1]) / (sum(gateset[2][1]) + sum(gateset[1][1]))
+        if num_qubits == 1:
+            return [
+                (GateInstruction((qubits[0],), i),)
+                for i in self._rng.choice(
+                    np.array(gateset[1][0], dtype=Instruction),
+                    p=[i / norm1q for i in gateset[1][1]],
+                    size=length,
+                )
+            ]
+
+        if not isinstance(self.coupling_map, List):
+            raise QiskitError("The coupling map must be set correctly before sampling.")
+
+        layer_list = []
+        for _ in list(range(length)):
+            all_edges = self.coupling_map[:]  # make copy of coupling map from which we pop edges
+            selected_edges = []
+            while all_edges:
```

Comment:
That code is neater, but doesn't it still need to be inside the `for _ in range(length):` loop to generate `selected_edges` a number of times equal to `length` so there will still be a nested loop? Or do you mean refactoring more so that this loop generates all needed instances of `selected_edges`?

Reply:
I benchmarked my code
```
707 µs ± 41.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
and current implementation
```
62.3 µs ± 1.67 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

likely initialization cost for `CouplingMap` is a bottleneck in my code. I was worried the list copy became a bottleneck since this loop was called for hundreds of times. Likely current implementation is not too bad. Let's go as is.